### PR TITLE
update checksum fn and assert nf test

### PR DESCRIPTION
**Pull request by zeeshanlakhani**
- [x] Waiting on one more test packet w/ a payload.

### CODE REVIEW

File: framework/src/headers/tcp.rs [hunk 1]
```diff
@@ -382,12 +383,27 @@ where
...
     }
 
+    /// Update Checksum w/ tcp_segment_length (from <Packet>) and src and dst on
+    /// header.
+    /// TODO: Validate Checksum?
+    pub fn update_checksum(&mut self, segment_length: usize, src: Ipv6Addr, dst: Ipv6Addr) {
+        self.set_checksum(0);
+
+        let mut sum = segment_length as u32
+            + src.segments().iter().map(|x| *x as u32).sum::<u32>()
```
Comment (awangc) [inline framework/src/headers/tcp.rs]: Nice! This `segments` function seems made for the checksum.

Comment (zeeshanlakhani) [reply framework/src/headers/tcp.rs]: yep, the only key was figuring out the right casting needs.

File: test/tcp-checksum/src/nf.rs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        .transform(box |pkt| {
+            let init_checksum = pkt.get_header().checksum();
+            let src = pkt.read_metadata().src_ip;
+            let dst = pkt.read_metadata().dst_ip;
+
+            println!(
+                "{}",
+                format!(
+                    "CheckSum: {} ~ {:x?} | Src: {} | Dst: {}",
+                    init_checksum, init_checksum, src, dst
```
[the remark this answers is not in the document]
Comment (zeeshanlakhani) [reply test/tcp-checksum/src/nf.rs]: one's printing the hex value yo.

File: framework/src/headers/tcp.rs [hunk 3]
```diff
@@ -382,12 +383,26 @@ where
...
+    /// TODO: Validate Checksum?
+    pub fn update_checksum(&mut self, segment_length: u32, src: Ipv6Addr, dst: Ipv6Addr) {
+        let mut sum = segment_length
+            + src.segments().iter().map(|x| *x as u32).sum::<u32>()
+            + dst.segments().iter().map(|x| *x as u32).sum::<u32>()
+            + TCP_NXT_HDR as u32;
+
+        while sum >> 16 != 0 {
+            sum = (sum >> 16) + (sum & 0xFFFF);
+        }
```
[the remark this answers is not in the document]
Comment (zeeshanlakhani) [reply framework/src/headers/tcp.rs]: yep. similar variations around, https://stackoverflow.com/questions/8845178/c-programming-tcp-checksum, but used the one from the a well-known packet impl in rust.

**Comment by zeeshanlakhani** [comment]
@clinedome 4 packets tested in the single pcap.

```shell
Running on node 0
Receiving started
Running 1 pipelines
Starting scheduler on 1
CheckSum: 44098 ~ ac42 | Src: fdaa:0:2:c::11 | Dst: fdaa:0:2:a::14
CheckSum: 44098 ~ ac42 | Src: fdaa:0:2:c::11 | Dst: fdaa:0:2:a::14
CheckSum: 34228 ~ 85b4 | Src: fdaa:0:2:c::11 | Dst: 2001:558:feed:10::2
CheckSum: 34228 ~ 85b4 | Src: fdaa:0:2:c::11 | Dst: 2001:558:feed:10::2
CheckSum: 42591 ~ a65f | Src: fdaa:0:1::13:0:0 | Dst: fdaa:0:1:c:0:11::
CheckSum: 42591 ~ a65f | Src: fdaa:0:1::13:0:0 | Dst: fdaa:0:1:c:0:11::
CheckSum: 8182 ~ 1ff6 | Src: 2001:558:feed:10:0:2:: | Dst: fdaa:0:1:c:0:11::
CheckSum: 8182 ~ 1ff6 | Src: 2001:558:feed:10:0:2:: | Dst: fdaa:0:1:c:0:11::
```

**Pull request merged**